src/encoder.rs

```diff
@@ -2303,44 +2309,63 @@ fn encode_partition_bottomup(seq: &Sequence, fi: &FrameInvariants, fs: &mut Fram
         let mode_decision = rdo_mode_decision(seq, fi, fs, cw, bsize, bo, spmvs, false).part_modes[0].clone();
 
         rd_cost = mode_decision.rd_cost + cost;
+        best_rd = rd_cost;
 
         encode_block_with_modes(seq, fi, fs, cw, w_pre_cdef, w_post_cdef, bsize, bo,
                             &mode_decision);
+
         best_decision = mode_decision;
+        best_pred_modes.push(best_decision.clone());
     }
 
     // Code a split partition and compare RD costs
     if can_split {
-        cw.rollback(&cw_checkpoint);
-        w_pre_cdef.rollback(&w_pre_checkpoint);
-        w_post_cdef.rollback(&w_post_checkpoint);
-
-        let nosplit_rd_cost = rd_cost;
-        partition = PartitionType::PARTITION_SPLIT;
-        {
-        subsize = bsize.subsize(partition);
-        let hbsw = subsize.width_mi(); // Half the block size width in blocks
-        let hbsh = subsize.height_mi(); // Half the block size height in blocks
+        for &partition in RAV1E_PARTITION_TYPES {
+            if partition == PartitionType::PARTITION_NONE {
+                continue;
+            }
+            assert!(bsw == bsh);
+            if must_split {
+                let cbw = (fi.w_in_b - bo.x).min(bsw); // clipped block width, i.e. having effective pixels
+                let cbh = (fi.h_in_b - bo.y).min(bsh);
+                let mut split_vert = false;
```

Comment:
My suggestion for this: set the value outside of the loop, near `can_split` and `must_split`, and call it `can_split_vert` (same with `split_horz`). It should simplify the content of the loop, which is quite complex now.

Reply:
I understand that, but would you still recommend moving those to outside the loop when those variables are only used if must_split is true?
The split_vert and split_horz are only used when must_split == true, when the input bsize straddles on the right or bottom frame borders, because there can be cases where horz and/or vert split is not possible so disallowed.

Reply:
OK, it does make more sense like this.